janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/AdjacentVertexOptimizerStrategy.java

```diff
@@ -0,0 +1,160 @@
+            P<?> predicate = parsePredicate(step);
+            replaceSequence(step, pos, predicate);
+        }
+    }
+
+    private boolean isOptimizableByPredicate(StepType step) {
+        return isValidPredicate(parsePredicate(step));
+    }
+
+    protected abstract P<?> parsePredicate(StepType step);
+
+    protected abstract boolean isValidPredicate(P<?> predicate);
+
+    private OptimizablePosition isOptimizableByPosition(StepType originalStep) {
+        Step<?, ?> predecessor = originalStep.getPreviousStep();
+        Step<?, ?> prePredecessor = predecessor.getPreviousStep();
+
+        // match predecessing out(), in() or both() steps
+        if (predecessor instanceof VertexStep<?>) {
+            if (((VertexStep<?>) predecessor).returnsVertex()) {
+                return OptimizablePosition.V2V_ID;
+            }
+            return OptimizablePosition.NONE;
+        }
+
+        // match predecessing inV(), outV() or otherV() steps
+        if (predecessor instanceof EdgeVertexStep || predecessor instanceof EdgeOtherVertexStep) {
+            if (prePredecessor instanceof VertexStep) {
+                // predecessor hast to operate on an edge type
+                if (((VertexStep<?>) prePredecessor).returnsEdge()) {
+                    return OptimizablePosition.V2E_E2V_ID;
+                }
+            }
+        }
+
+        return OptimizablePosition.NONE;
+    }
+
+    private void replaceSequence(StepType originalStep, OptimizablePosition pos,
+                                   P<?> predicate) {
+        switch (pos) {
+        case V2E_E2V_ID:
+            replaceSequence_V2E_E2V_ID(originalStep, predicate);
+            break;
+        case V2V_ID:
+            replaceSequence_V2V_ID(originalStep, predicate);
+            break;
+        default:
+            break;
+        }
+    }
+
+    private void replaceSequence_V2E_E2V_ID(StepType originalStep, P<?> predicate) {
+        Traversal.Admin<?,?> traversal = originalStep.getTraversal();
+
+        Step<Edge,Vertex> e2vStep;
+        if (originalStep.getPreviousStep() instanceof EdgeVertexStep) {
+            e2vStep = (EdgeVertexStep) originalStep.getPreviousStep();
+        } else if (originalStep.getPreviousStep() instanceof EdgeOtherVertexStep) {
+            e2vStep = (EdgeOtherVertexStep) originalStep.getPreviousStep();
+        } else {
+            return;
+        }
+
+        // create new has("~adjacent", id_value) step before e2v step
+        HasContainer hc = new HasContainer(ADJACENT_ID.name(), P.eq(predicate.getValue()));
+        HasStep<Edge> hasAdjacentIdStep = new HasStep<Edge>(traversal, hc);
+
+        // remove original selection step
+        traversal.removeStep(originalStep);
+
+        // insert new step
+        TraversalHelper.insertBeforeStep(hasAdjacentIdStep, e2vStep, traversal);
+    }
+
+    private void replaceSequence_V2V_ID(StepType originalStep, P<?> predicate) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The instance method name 'replaceSequence_V2V_ID' doesn't match '[a-z][a-zA-Z0-9]*'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=5005131)